src/DIRAC/Resources/IdProvider/IdProviderFactory.py

```diff
@@ -10,52 +10,111 @@
 from __future__ import division
 from __future__ import print_function
 
+import jwt
+
 from DIRAC import S_OK, S_ERROR, gLogger
-from DIRAC.Core.Utilities.ObjectLoader import ObjectLoader
-from DIRAC.ConfigurationSystem.Client.Helpers.Resources import getInfoAboutProviders
+from DIRAC.Core.Utilities import ObjectLoader, ThreadSafe
+from DIRAC.Core.Utilities.DictCache import DictCache
+from DIRAC.Resources.IdProvider.Utilities import getProviderInfo, getSettingsNamesForIdPIssuer
+from DIRAC.ConfigurationSystem.Client.Utilities import getAuthorizationServerMetadata
+from DIRAC.FrameworkSystem.private.authorization.utils.Clients import DEFAULT_CLIENTS
 
 __RCSID__ = "$Id$"
 
+gCacheMetadata = ThreadSafe.Synchronizer()
+
 
 class IdProviderFactory(object):
 
-  #############################################################################
   def __init__(self):
     """ Standard constructor
     """
     self.log = gLogger.getSubLogger('IdProviderFactory')
+    self.cacheMetadata = DictCache()
+
+  @gCacheMetadata
+  def getMetadata(self, idP):
+    return self.cacheMetadata.get(idP) or {}
 
-  #############################################################################
-  def getIdProvider(self, idProvider):
-    """ This method returns a IdProvider instance corresponding to the supplied name.
+  @gCacheMetadata
+  def addMetadata(self, idP, data, time=24 * 3600):
+    if data:
+      self.cacheMetadata.add(idP, time, data)
 
-        :param str idProvider: the name of the Identity Provider
+  def getIdProviderForToken(self, token):
+    """ This method returns a IdProvider instance corresponding to the supplied
+        issuer in a token.
+
+        :param token: access token or dict with access_token key
 
         :return: S_OK(IdProvider)/S_ERROR()
     """
-    result = getInfoAboutProviders(of='Id', providerName=idProvider, option="all", section="all")
+    if isinstance(token, dict):
+      token = token['access_token']
+
+    data = {}
+
+    # Read token without verification to get issuer
+    issuer = jwt.decode(token, leeway=300,
+                        options=dict(verify_signature=False, verify_aud=False))['iss'].strip('/')
+
+    result = getSettingsNamesForIdPIssuer(issuer)
+    if result['OK']:
+      return self.getIdProvider(result['Value'][0])
+
+    _result = getAuthorizationServerMetadata()
+    if not _result['OK']:
+      return _result
+    if issuer == _result['Value'].get('issuer', '').strip('/'):
+      return self.getIdProvider(DEFAULT_CLIENTS.keys()[0])
```

Comment:
that's python2 only

Reply:
It's also not well defined in Python 2 as the dictionary order isn't predictable.

Reply:
fixed